### PR TITLE
Model Tag Roller

### PR DESCRIPTION
**Pull request by WispyMouse**
- GetAllTags added to Model and its implementations
- Roller can report on tags

### CODE REVIEW

File: ParquetClassLibrary/Biomes/BiomeModel.cs [hunk 1]
```diff
@@ -64,5 +64,16 @@ public sealed class BiomeModel : Model
             EntryRequirements = (inEntryRequirements ?? Enumerable.Empty<ModelTag>()).ToList();
         }
         #endregion
+
+        #region Utilities
+        /// <summary>
+        /// Returns a collection of all <see cref="ModelTag"/>s the <see cref="Model"/> has applied to it. Classes inheriting from <see cref="Model"/> that include <see cref="ModelTag"/> should override accordingly.
```
Comment (mxashlynn) [inline ParquetClassLibrary/Biomes/BiomeModel.cs:70]: I love this way of handling it.

File: ParquetClassLibrary/Biomes/BiomeModel.cs [hunk 2]
```diff
@@ -64,5 +64,16 @@ public sealed class BiomeModel : Model
...
+        /// </summary>
+        /// <returns>List of all <see cref="ModelTag"/>s.</returns>
+        public override IEnumerable<ModelTag> GetAllTags()
+        {
+            return base.GetAllTags().Union(ParquetCriteria).Union(EntryRequirements);
```
Comment (mxashlynn) [inline ParquetClassLibrary/Biomes/BiomeModel.cs:75]: And I love this way of implementing it!
Great work!

File: ParquetRoller/Roller.cs [hunk 3]
```diff
@@ -413,9 +413,31 @@ private static ExitCode ListMaxIDs(ModelCollection inWorkload)
...
+                foreach (var modelTag in model.GetAllTags())
+                {
+                    if (!allTags.Any(tag => tag.CompareTo(modelTag) == 0))
+                    {
+                        allTags.Add(modelTag);
+                    }
+                }
+            }
+
+            foreach (var modelTag in allTags.Distinct())
```
Comment (mxashlynn) [inline ParquetRoller/Roller.cs:435]: Given that allTags is a HashSet and that you only add tags that do not compare as equivalent, shouldn't these already be distinct?

**Review by mxashlynn — CHANGES_REQUESTED** [review]
Excellent work.
I did have one small question before we merge, but overall I am impressed by how you handled this one!

File: ParquetRoller/Roller.cs (hunk 3 above)
Comment (WispyMouse) [reply ParquetRoller/Roller.cs:435]: Oh, true! Removed.

**Pull request closed without merging**